src/target/riscv/riscv-013.c

```diff
@@ -1273,192 +1306,228 @@ static bool has_sufficient_progbuf(struct target *target, unsigned size)
-			LOG_ERROR("Unsupported register (enum gdb_regno)(%d)", number);
-			return ERROR_FAIL;
+			return internal_register_read64_progbuf_scratch(target, &program,
+					value);
 		}
-		if (register_write_direct(target, GDB_REGNO_S0, value) != ERROR_OK)
+		if (riscv_program_insert(&program,
+					riscv_supports_extension(target, 'D') ?
+					fmv_x_d(S0, freg) : fmv_x_w(S0, freg)) != ERROR_OK)
 			return ERROR_FAIL;
+	} else if (number >= GDB_REGNO_CSR0 && number <= GDB_REGNO_CSR4095) {
+		if (riscv_program_csrr(&program, S0, number) != ERROR_OK)
+			return ERROR_FAIL;
+	} else {
+		LOG_ERROR("Unsupported register: %s", gdb_regno_name(number));
+		return ERROR_FAIL;
 	}
 
-	int exec_out = riscv_program_exec(&program, target);
-	/* Don't message on error. Probably the register doesn't exist. */
-	if (exec_out == ERROR_OK && target->reg_cache) {
-		struct reg *reg = &target->reg_cache->reg_list[number];
-		buf_set_u64(reg->value, 0, reg->size, value);
-	}
+	if (riscv_program_exec(&program, target) != ERROR_OK)
+		return ERROR_FAIL;
 
-	if (use_scratch)
-		scratch_release(target, &scratch);
+	return register_read_abstract(target, value, GDB_REGNO_S0) != ERROR_OK;
+}
 
-	if (cleanup_after_register_access(target, mstatus, number) != ERROR_OK)
+/**
+ * This function is used to write a 64-bit value to a register by executing a
+ * program.
+ * The program loads a value from address located in S0 to a register.
+ * The caller should save S0.
+ */
+static int internal_register_write64_progbuf_scratch(struct target *target,
+		struct riscv_program *program, riscv_reg_t value)
+{
+	scratch_mem_t scratch;
+
+	if (scratch_reserve(target, &scratch, program, 8) != ERROR_OK)
 		return ERROR_FAIL;
 
-	return exec_out;
+	if (register_write_abstract(target, GDB_REGNO_S0, scratch.hart_address)
+			!= ERROR_OK) {
+		scratch_release(target, &scratch);
+		return ERROR_FAIL;
+	}
+	if (scratch_write64(target, &scratch, value) != ERROR_OK) {
+		scratch_release(target, &scratch);
+		return ERROR_FAIL;
+	}
+	int result = riscv_program_exec(program, target);
+
+	scratch_release(target, &scratch);
+	return result;
 }
 
-/** Actually read registers from the target right now. */
-static int register_read_direct(struct target *target, uint64_t *value, uint32_t number)
+/**
+ * This function writes a register by writing a program to program buffer and
+ * executing it.
+ */
+static int register_write_progbuf(struct target *target, enum gdb_regno number,
+		riscv_reg_t value)
 {
-	if (dm013_select_target(target) != ERROR_OK)
+	assert(target->state == TARGET_HALTED);
+
+	struct riscv_program program;
+
+	riscv_program_init(&program, target);
+
+	if (riscv_save_register(target, GDB_REGNO_S0) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (number >= GDB_REGNO_FPR0 && number <= GDB_REGNO_FPR31 &&
+			riscv_supports_extension(target, 'D') && riscv_xlen(target) < 64) {
+		/* There are no instructions to move all the bits from a register,
+		 * so we need to use some scratch RAM.
+		 */
```

Comment:
Detail: No need to have the closing `*/` on a new line.

```suggestion
		/* There are no instructions to move all the bits from a register,
		 * so we need to use some scratch RAM. */
```

Reply:
When I try to delete this newline there is a warning from `checkpatch`:
```
git diff --patch github/en-sc/optimize-access | filterdiff -x ... | ~/openocd/tools/scripts/checkpatch.pl --no-signoff -
WARNING: Block comments use a trailing */ on a separate line
#18: FILE: src/target/riscv/riscv-013.c:1431:
+                * so we need to use some scratch RAM.*/

total: 0 errors, 1 warnings, 20 lines checked
...
```
If you insist, I'm open to accepting your suggestion, but shouldn't then `.github/workflows/checkpatch.yml` and/or `tools/scripts/checkpatch.pl` be adjusted?

Reply:
Ah, I didn't know `checkpatch` requires this style.

What if you remove the start at the beginning of the line like shown below? Will `checkpatch` still complain?

```suggestion
		/* There are no instructions to move all the bits from a register,
		   so we need to use some scratch RAM. */
```

Reply:
Yes, It will
```
> git diff github/en-sc/optimize-access
...
                /* There are no instructions to move all the bits from a register,
-                * so we need to use some scratch RAM.
-                */
+                  so we need to use some scratch RAM. */
...
>  git diff --patch github/en-sc/optimize-access | filterdiff -x ... | ~/openocd/tools/scripts/checkpatch.pl --no-signoff -
WARNING: Block comments use a trailing */ on a separate line
#18: FILE: src/target/riscv/riscv-013.c:1429:
+                  so we need to use some scratch RAM. */
...
```

Reply:
Ok, thanks for checking it. So let's leave the comments as they are now.